crates/dwallet-mpc-types/src/dwallet_mpc.rs

```diff
@@ -31,6 +32,99 @@ pub const START_NETWORK_DKG_EVENT_STRUCT_NAME: &IdentStr =
+    /// A fragment of the original message.
+    pub message: MPCMessage,
+    /// The position of this chunk in the original message sequence.
+    pub sequence_number: u64,
+    /// Total number of chunks in the message.
+    pub number_of_chunks: usize,
+}
+
+/// Represents the state of the message building process.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub enum MessageState {
+    /// All chunks received; full message reconstructed.
+    Complete(MPCMessage),
+    /// Still waiting on some chunks; maps sequence numbers to slices.
+    Incomplete(HashMap<u64, MPCMessageSlice>),
+}
+
+/// Builds and reconstructs messages from incoming slices.
+/// Some MPC messages might be greater than the maximum size of a consensus message limit.
+/// The `MPCMessageBuilder` is used to split the message into smaller chunks,
+/// to avoid exceeding the maximum size limit.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct MPCMessageBuilder {
+    /// Current state of the message.
+    pub messages: MessageState,
+}
+
+impl MPCMessageBuilder {
+    /// Splits a message into smaller chunks and returns a new builder with those chunks.
+    /// Ensures at least one chunk is created, even if the message is empty.
+    pub fn split(message: Vec<u8>, chunk_size: usize) -> Self {
+        let chunks: Vec<Vec<u8>> = if message.is_empty() {
+            vec![vec![]] // Ensure at least one slice for empty messages
+        } else {
+            message
+                .chunks(chunk_size)
+                .map(|chunk| chunk.to_vec())
+                .collect()
+        };
+
+        let number_of_chunks = chunks.len();
+
+        let messages = chunks
+            .into_iter()
+            .enumerate()
+            .map(|(i, message)| {
+                (
+                    i as u64,
+                    MPCMessageSlice {
+                        message,
+                        sequence_number: i as u64,
+                        number_of_chunks,
+                    },
+                )
+            })
+            .collect();
+
+        Self {
+            messages: MessageState::Incomplete(messages),
+        }
+    }
+
+    /// Adds a single message slice to the builder.
+    pub fn add_message(&mut self, message: MPCMessageSlice) {
+        if let MessageState::Incomplete(messages) = &mut self.messages {
+            messages.insert(message.sequence_number, message);
+        }
+    }
+
+    /// Attempts to build the full message from all available slices.
+    /// Converts state to `Complete` if all slices are present.
+    pub fn build_message(&mut self) {
+        if let MessageState::Incomplete(messages) = &self.messages {
+            if let Some(expected_chunks) = messages.values().next().map(|s| s.number_of_chunks) {
+                if messages.len() == expected_chunks {
+                    let mut message = Vec::new();
+
+                    for i in 0..expected_chunks {
```

Comment:
Functionally written, this would look like this `0..number_of_chunks).map(|i| messages.get(&i)).reduce(...)`

The code written in this repo is needlessly iterative, it could have easily been written functionally which is how Rust code should be